crypto/provider_core.c

```diff
@@ -339,20 +428,54 @@ int ossl_provider_forall_loaded(OPENSSL_CTX *ctx,
     struct provider_store_st *store = get_provider_store(ctx);
 
     if (store != NULL) {
+        int found_activated = 0;
+
         CRYPTO_THREAD_read_lock(store->lock);
-        for (i = 0; i < sk_OSSL_PROVIDER_num(store->providers); i++) {
-            OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(store->providers, i);
+        ret = provider_forall_loaded(store, &found_activated, cb, cbdata);
 
-            if (prov->flag_initialized
-                && !(ret = cb(prov, cbdata)))
-                break;
+        /*
+         * If there's nothing activated ever in this store, try to activate
+         * all fallbacks.
+         */
+        if (!found_activated && store->use_fallbacks) {
+            int num_provs = sk_OSSL_PROVIDER_num(store->providers);
+
+            for (i = 0; i < num_provs; i++) {
+                OSSL_PROVIDER *prov =
+                    sk_OSSL_PROVIDER_value(store->providers, i);
+
+                /*
+                 * Note that we don't care if the activation succeeds or
+                 * not.  If it doesn't succeed, then the next loop will
+                 * fail anyway.
+                 */
+                if (prov->flag_fallback)
+                    provider_activate(prov);
+            }
+
+            /*
+             * Now that we've activated available fallbacks, try a second sweep
+             */
+            ret = provider_forall_loaded(store, NULL, cb, cbdata);
+
+            /* TODO: should we set store->use_fallbacks = 0 here? */
```

Comment:
IMO - yes we should. IIUC, it should normally be 0 here anyway (because we will normally always have at least one fallback). The only way we should get here is if we have already attempted to active fallbacks, but we still didn't end up activating any providers. In that case it makes sense to set use_fallbacks to 0 anyway because doing it again seems unlikely to succeed a second time.

Reply:
I'm imagining the following scenario:

1. an algorithm is fetched, and since no provider is activated, "default" gets activated as a fallback.
2. another fallback provider is added (let's call it "foo").
3. an algorithm is fetched, isn't found among current providers (i.e. "default", as that's the only one activated so far), but would be found if "foo" was activated.

Of course, that scenario isn't possible if we decide that *all* fallbacks must be in place before algorithm fetches are done.

Reply:
Surely in that scenario `use_fallbacks` would get set to 0 as soon as `default` got activated in step 1. Therefore we never attempt to activate fallbacks again after that point so we would never attempt to active "foo". IMO fallbacks are mostly only interesting in scenarios where applications are not provider aware. As soon as an application start dynamically adding and removing providers then we assume it knows what its doing and fallbacks shouldn't be necessary.

Reply:
Note that fallbacks can only be set internally...  I guess I'm just try to think forward of possible corner cases...  But you know what, I can always set use_fallbacks to zero here with an added comment that we might want to reconsider it.